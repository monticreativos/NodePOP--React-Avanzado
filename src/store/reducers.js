import {
  AUTH_LOGIN_FAILURE,
  AUTH_LOGIN_REQUEST,
  AUTH_LOGIN_SUCCESS,
  AUTH_LOGOUT_SUCCESS,
  ADVERT_CREATED_SUCCESS,
  ADVERTS_LOADED_SUCCESS,
  ADVERTS_LOADED_FAILURE,
  UI_RESET_ERROR,
  ADVERT_LOADED_SUCCESS,
  ADVERT_LOADED_REQUEST,
  ADVERT_LOADED_FAILURE,
  ADVERT_CREATED_REQUEST,
  ADVERTS_LOADED_REQUEST,
  ADVERT_CREATED_FAILURE,
  ADVERTS_FILTER_REQUEST,
  ADVERTS_FILTER_SUCCESS,
  ADVERTS_FILTER_FAILURE,
  ACCESS_TOKEN,
  ADVERT_TAGS,
  ADVERT_DELETED_SUCCESS,
  ADVERT_DELETED_FAILURE,
  TAGS_LOADED,
} from './types'

export const defaultState = {
  auth: false,
  adverts: {
    loaded: false,
    data: [],
  },
  tags: [],
  ui: {
    isLoading: false,
    error: null,
  },
}

export const auth = (state = defaultState, action) => {
  switch (action.type) {
    case AUTH_LOGIN_SUCCESS:
      console.log(action.payload)

      return { ...state, auth: true, token: action.payload }
    case AUTH_LOGOUT_SUCCESS:
      return false
    case ACCESS_TOKEN:
      return state.auth.token
    default:
      return state
  }
}

export const adverts = (state = defaultState.adverts, action) => {

  switch (action.type) {
    case ADVERTS_LOADED_SUCCESS:
      return { loaded: true, data: action.payload }
    case ADVERT_LOADED_SUCCESS:
      return { ...state, data: [...state.data, action.payload] }
    case ADVERT_CREATED_SUCCESS:
      return { ...state, data: [action.payload, ...state.data] }
    case ADVERT_TAGS:
      return { ...state, tags: action.payload }
    case ADVERTS_FILTER_SUCCESS:
      return { ...state, data: [action.payload] }
    case ADVERT_DELETED_SUCCESS:
      return {
        ...state,
        data: state.data.filter(advert => advert.id !== action.payload),
      };
    default:
      return state
  }
}

export const ui = (state = defaultState.ui, action) => {
  switch (action.type) {
    case AUTH_LOGIN_REQUEST:
    case ADVERTS_LOADED_REQUEST:
    case ADVERT_LOADED_REQUEST:
    case ADVERTS_FILTER_REQUEST:
    case ADVERT_CREATED_REQUEST:
      return { ...state, isLoading: true, error: null }
    case AUTH_LOGIN_SUCCESS:
    case ADVERTS_LOADED_SUCCESS:
    case ADVERT_LOADED_SUCCESS:
    case ADVERT_CREATED_SUCCESS:
      return { ...state, isLoading: false }
    case AUTH_LOGIN_FAILURE:
    case ADVERTS_LOADED_FAILURE:
    case ADVERT_LOADED_FAILURE:
    case ADVERTS_FILTER_FAILURE:
    case ADVERT_CREATED_FAILURE:
    case ADVERT_DELETED_FAILURE:
      return { ...state, isLoading: false, error: action.payload }
    case UI_RESET_ERROR:
      return { ...state, error: null }
    default:
      return state
  }
}
export const tags = (state = defaultState.tags, action) =>
  action.type === TAGS_LOADED ? action.payload : state;

