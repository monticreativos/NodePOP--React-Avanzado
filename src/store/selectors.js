export const getIsLogged = (state) => state.auth

export const getToken = (state) => state.auth.token

export const getAdverts = (state) => state.adverts.data

export const getAreAdvertsLoaded = (state) => state.adverts.loaded

export const getAdvert = (state, advertId) =>
  getAdverts(state).find(advert => advert.id === advertId);


export const getUi = (state) => state.ui

export const getDeletedAdvert = (advertId) => (state) =>
  state.adverts.data.filter((advert) => advert.id != advertId)
