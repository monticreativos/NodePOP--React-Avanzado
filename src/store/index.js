import { createStore, combineReducers, applyMiddleware } from 'redux'
import * as reducers from './reducers'
import { composeWithDevTools } from '@redux-devtools/extension'
import thunk from 'redux-thunk'

import * as auth from '../components/auth/service'
import * as adverts from '../components/adverts/service'
import {
  AUTH_LOGIN_SUCCESS,
  HISTORY_BACK,
  ADVERT_CREATED_SUCCESS,
} from './types'

const api = { auth, adverts }

function logger(store) {
  return function (next) {
    return function (action) {
      console.log('before action', action, store.getState())
      const result = next(action)
      console.log('after action', action, store.getState())
      return result
    }
  }
}

const timestamp = () => (next) => (action) => {
  const newAction = {
    ...action,
    meta: {
      ...action.meta,
      timestamp: new Date(),
    },
  }
  return next(newAction)
}

const allState = (store) => (next) => (action) => {
  const newAction = {
    ...action,
    meta: {
      ...action.meta,
      state: store.getState(),
    },
  }
  return next(newAction)
}

const failureRedirections = (history, redirections) => (_store) => (next) => (
  action,
) => {
  const result = next(action)
  if (action.error) {
    const redirection = redirections[action.payload.status]
    if (redirection) {
      history.push(redirection)
    }
  }

  return result
}

const successRedirections = (history, redirections) => (_store) => (next) => (
  action,
) => {
  const result = next(action)
  const redirection = redirections[action.type]
  if (redirection) {
    redirection(history, action.payload)
  }

  return result
}

const rootReducer = combineReducers(reducers)

const historyHighOrderReducer = (rootReducer) => {
  return (state, action) => {
    const { history, ...rootState } = state
    if (action.type === HISTORY_BACK) {
      const newHistory = history.slice(0, history.length - 1)
      return {
        ...newHistory[newHistory.length - 1].state,
        history: newHistory,
      }
    }
    const newState = rootReducer(rootState, action)
    return {
      ...newState,
      history: [...(history || []), { action, state: newState }],
    }
  }
}

const configureStore = (preloadedState, { history }) => {
  const middlewares = [
    thunk.withExtraArgument({ api, history }),
    failureRedirections(history, { 401: '/login', 404: '/404' }),
    successRedirections(history, {
      [AUTH_LOGIN_SUCCESS]: (history) => {
        const from = history.location.state?.from?.pathname || '/'
        history.replace(from)
      },
      [ADVERT_CREATED_SUCCESS]: (history, payload) =>
        history.push(`/adverts/${payload.id}`),
    }),
    logger,
    timestamp,
    // allState,
  ]

  const store = createStore(
    historyHighOrderReducer(rootReducer),
    preloadedState,
    composeWithDevTools(applyMiddleware(...middlewares)),
  )

  return store
}

export default configureStore
