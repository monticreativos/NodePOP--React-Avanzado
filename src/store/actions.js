import { getAreAdvertsLoaded, getAdvert, getAreTagsLoaded } from './selectors'
import {
  AUTH_LOGIN_FAILURE,
  AUTH_LOGIN_REQUEST,
  AUTH_LOGIN_SUCCESS,
  AUTH_LOGOUT_SUCCESS,
  ADVERTS_LOADED_FAILURE,
  ADVERTS_LOADED_REQUEST,
  ADVERTS_LOADED_SUCCESS,
  ADVERT_CREATED_FAILURE,
  ADVERT_CREATED_SUCCESS,
  ADVERT_LOADED_FAILURE,
  ADVERT_LOADED_SUCCESS,
  UI_RESET_ERROR,
  ADVERT_DELETED_SUCCESS,
  ADVERT_DELETED_FAILURE,
  ADVERT_LOADED_REQUEST,
  TAGS_LOADED,
} from './types'


// Creamos los disparadores hacia el redux
export const authLoginRequest = () => ({
  type: AUTH_LOGIN_REQUEST,
})
export const authLoginSuccess = (token) => ({
  type: AUTH_LOGIN_SUCCESS,
  payload: token,
})

export const authLoginFailure = (error) => ({
  type: AUTH_LOGIN_FAILURE,
  payload: error,
  error: true,
})


// Auth Login action
export const authLogin = (credentials) => {
  return async function (dispatch, _getState, { api, history }) {
    dispatch(authLoginRequest())
    try {
      const response = await api.auth.login(credentials)
      console.log(response)
      dispatch(authLoginSuccess(response))
      // redirect...
      const from = history.location.state?.from?.pathname || '/adverts'
      console.log(from)
      history.replace(from)
    } catch (error) {
      dispatch(authLoginFailure(error))
    }
  }
}



// Creamos los disparadores hacia el redux
export const authLogoutSuccess = () => ({
  type: AUTH_LOGOUT_SUCCESS,
})


// Auth Logout action
export const authLogout = () => {
  return function (dispatch, _getState, { api }) {
    return api.auth.logout().then(() => {
      dispatch(authLogoutSuccess())
    })
  }
}



// Creamos los disparadores hacia el redux
export const advertsLoadedRequest = () => ({
  type: ADVERTS_LOADED_REQUEST,
})

export const advertsLoadedSuccess = (adverts) => ({
  type: ADVERTS_LOADED_SUCCESS,
  payload: adverts,
})

export const advertsLoadedFailure = (error) => ({
  type: ADVERTS_LOADED_FAILURE,
  payload: error,
  error: true,
})


// Adverts Loaded action
export const advertsLoaded = () => {
  return async function (dispatch, getState, { api }) {
    const advertssLoaded = getAreAdvertsLoaded(getState())
    if (advertssLoaded) return

    dispatch(advertsLoadedRequest())
    try {
      const adverts = await api.adverts.getAdverts()
      dispatch(advertsLoadedSuccess(adverts))
      return
    } catch (error) {
      dispatch(advertsLoadedFailure(error))
    }
  }
}




// Creamos los disparadores hacia el redux
export const advertLoadedRequest = () => ({
  type: ADVERT_LOADED_REQUEST,
});

export const advertLoadedSuccess = (advert) => ({
  type: ADVERT_LOADED_SUCCESS,
  payload: advert,
})

export const advertLoadedFailure = (error) => ({
  type: ADVERT_LOADED_FAILURE,
  payload: error,
  error: true,
})


// Advert Loaded by ID action
export const advertLoaded = (advertId) => {
  return async function (dispatch, getState, { api }) {
    if (getAdvert(getState(), advertId)) {
      return;
    }
    dispatch(advertLoadedRequest());
    try {
      const advert = await api.adverts.getAdvert(advertId);
      dispatch(advertLoadedSuccess(advert));
    } catch (error) {
      dispatch(advertLoadedFailure(error));
    }
  };
}



// Creamos los disparadores hacia el redux
export const advertCreatedSuccess = (advert) => ({
  type: ADVERT_CREATED_SUCCESS,
  payload: advert,
})

export const advertCreatedFailure = (error) => ({
  type: ADVERT_CREATED_FAILURE,
  payload: error,
  error: true,
})


// Advert Created action
export const advertCreated = (advert) => {
  return async function (dispatch, _getState, { api, history }) {
    // dispatch(tweetCreatedRequest());
    try {
      const { id } = await api.adverts.createAdvert(advert)
      const createdAdvert = await api.adverts.getAdvert(id)
      dispatch(advertCreatedSuccess(createdAdvert))
      return createdAdvert
    } catch (error) {
      dispatch(advertCreatedFailure(error))
    }
  }
}



// Creamos los disparadores hacia el redux
export const advertDeletedSuccess = (adverts) => ({
  type: ADVERT_DELETED_SUCCESS,
  payload: adverts,
})

export const advertDeletedFailure = (error) => ({
  type: ADVERT_DELETED_FAILURE,
  payload: error,
  error: true,
})


// Advert Deleted by ID action
export const advertDeleted = (id) => {
  return async function (dispatch, getState, { api, history }) {
    // dispatch(tweetCreatedRequest());
    try {
      await api.adverts.deleteAdvert(id)
        dispatch(advertDeletedSuccess(id))
        history.push(`/adverts`);
    } catch (error) {
      dispatch(advertDeletedFailure(error))
    }
  }
}



// Creamos los disparadores hacia el redux
export const uiResetError = () => ({
  type: UI_RESET_ERROR,
})

// Creamos los disparadores hacia el redux
export const tagsLoaded = tags => ({
  type: TAGS_LOADED,
  payload: tags,
});


// Load Tags action
export const loadTags = () => {
  return async function (dispatch, getState, { api }) {
    if (getAreTagsLoaded(getState())) {
      return;
    }
    const tags = await api.adverts.getTags();
    dispatch(tagsLoaded(tags));
  };
};
