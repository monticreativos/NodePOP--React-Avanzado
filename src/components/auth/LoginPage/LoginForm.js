import T from 'prop-types'
import { useDispatch, useSelector } from 'react-redux'
import { authLogin, uiResetError } from '../../../store/actions'

import useForm from '../../../hooks/useForm'

const validEmail = ({ email }) => email
const validPassword = ({ password }) => password

function LoginForm({ onSubmit }) {
  const dispatch = useDispatch()
  const {
    formValue: credentials,
    handleChange,
    // handleSubmit,
    validate,
  } = useForm({
    email: '',
    password: '',
    remember: false,
  })
  const { email, password, remember } = credentials

  const handleSubmit = async (event) => {
    // event.preventDefault()
    dispatch(authLogin(credentials))
  }

  return (
    <form onSubmit={handleSubmit}>
      <input name="email" value={email} onChange={handleChange} />
      <input
        type="password"
        name="password"
        value={password}
        onChange={handleChange}
      />
      <input
        type="checkbox"
        name="remember"
        checked={remember}
        onChange={handleChange}
      />
      <button disabled={!validate(validEmail, validPassword)}>Login</button>
    </form>
  )
}

LoginForm.propTypes = {
  onSubmit: T.func.isRequired,
}

export default LoginForm
