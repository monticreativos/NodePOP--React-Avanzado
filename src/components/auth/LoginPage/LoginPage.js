import { useCallback, useEffect, useMemo, useRef, useState } from 'react'
import Button from '../common/Button'
import FormField from '../common/FormField'
import T from 'prop-types'
import { getIsLogged } from '../../../store/selectors'
import { Navigate } from 'react-router-dom'

import './LoginPage.css'
import { useDispatch, useSelector } from 'react-redux'
import { authLogin, uiResetError } from '../../../store/actions'
import { getUi } from '../../../store/selectors'

function useRenders() {

  const count = useRef(1)
  const isLogged = useSelector(getIsLogged)
  console.log(isLogged)

  useEffect(() => {
    count.current++
  })

  if (isLogged) {
    return <Navigate to="/adverts" replace />
  }

  return count.current
}

function LoginPage() {
  const renders = useRenders()
  const ref = useRef(null)
  const [credentials, setCredentials] = useState({
    email: '',
    password: '',
    remember: false,
  })
  const dispatch = useDispatch()
  const { isLoading, error } = useSelector(getUi)

  useEffect(() => {
    console.log(ref.current)
    ref.current.focus()
  }, [])

  const { email, password, remember } = credentials

  const handleChange = useCallback(
    ({ target: { value, name, type, checked } }) => {
      setCredentials((credentials) => ({
        ...credentials,
        [name]: type === 'checkbox' ? checked : value,
      }))
    },
    [],
  )

  const resetError = () => dispatch(uiResetError())

  const handleSubmit = async (event) => {
    event.preventDefault()
    dispatch(authLogin(credentials))
  }

  const buttonDisabled = useMemo(() => {
    console.log('calculando...')
    return !email || !password || isLoading
  }, [email, password, isLoading])

  return (
    <div className="loginPage">
      {renders}
      <h1 className="loginPage-title">Log in to Nodepop</h1>
      <form className="loginForm" onSubmit={handleSubmit}>
        <FormField
          type="text"
          name="email"
          label="phone, email or username"
          className="loginForm-field"
          value={email}
          onChange={handleChange}
          // ref={ref}
        />
        <FormField
          type="password"
          name="password"
          label="password"
          className="loginForm-field"
          value={password}
          onChange={handleChange}
          ref={ref}
        />
        <input
          type="checkbox"
          name="remember"
          checked={remember}
          value="remember"
          onChange={handleChange}
        />
        <select value="2" onChange={(event) => console.log(event)}>
          <option value="1">Option 1</option>
          <option value="2">Option 2</option>
          <option value="3">Option 3</option>
        </select>
        <input
          type="file"
          onChange={(event) => console.log(event.target.files[0])}
        />

        <Button
          className="loginForm-submit"
          type="submit"
          variant="primary"
          disabled={buttonDisabled}
        >
          Log in
        </Button>
      </form>
      {error && (
        <div onClick={resetError} className="loginPage-error">
          {error.message}
        </div>
      )}
    </div>
  )
}

LoginPage.propTypes = {
  onLogin: T.func,
}

export default LoginPage
