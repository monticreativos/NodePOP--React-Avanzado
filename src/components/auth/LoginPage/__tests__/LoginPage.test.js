import { fireEvent, render, screen } from '@testing-library/react'
import { Provider } from 'react-redux'
import LoginPage from '../LoginPage'
import { authLogin, uiResetError } from '../../../../store/actions'
import { BrowserRouter as Router } from 'react-router-dom'
import '@testing-library/jest-dom';

jest.mock('../../../../store/actions')

describe('LoginPage', () => {
  test('snapshot', () => {
    const state = {
      ui: {
        error: null,
        isLoading: false,
      },
    }
    const store = {
      getState: () => state,
      dispatch: jest.fn(),
      subscribe: () => {},
    }
    const { container } = render(
      <Router>
        <Provider store={store}>
          <LoginPage />
        </Provider>
      </Router>,
    )
    expect(container).toMatchSnapshot()
  })

  test('should authLogin action', () => {
    authLogin.mockReturnValue('action')
    const state = {
      ui: {
        error: null,
        isLoading: false,
      },
    }
    const username = '1234@admin.com'
    const password = '1234'
    const store = {
      getState: () => state,
      dispatch: jest.fn(),
      subscribe: () => {},
    }
    render(
      <Provider store={store}>
        <LoginPage />
      </Provider>,
    )
    const usernameInput = screen.getByLabelText(/username/)
    const passwordInput = screen.getByLabelText(/password/)
    const submitButton = screen.getByRole('button')

    expect(submitButton).toBeDisabled
    fireEvent.change(usernameInput, { target: { value: username } })
    fireEvent.change(passwordInput, { target: { value: password } })
    expect(submitButton).not.toBeDisabled

    // fireEvent.click(submitButton)
    // const credentials = authLogin.mock.calls[0][0]
    // expect(credentials).toMatchObject({ username, password })
    // expect(store.dispatch).toHaveBeenCalledWith('action')
  })

  //   test('should reset error', () => {
  //     const error = { message: 'Network error' }
  //     const state = {
  //       ui: {
  //         error,
  //         isLoading: false,
  //       },
  //     }
  //     const store = {
  //       getState: () => state,
  //       dispatch: jest.fn(),
  //       subscribe: () => {},
  //     }
  //     render(
  //       <Provider store={store}>
  //         <LoginPage />
  //       </Provider>,
  //     )
  //     const errorElement = screen.getByText(error.message)
  //     expect(errorElement).toBeInTheDocument()
  //     expect(screen.queryByText('not exist')).not.toBeInTheDocument()
  //     fireEvent.click(errorElement)
  //     expect(uiResetError).toHaveBeenCalled()
  //   })
})
