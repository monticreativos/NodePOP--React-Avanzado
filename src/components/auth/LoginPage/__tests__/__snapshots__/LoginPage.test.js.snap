// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage snapshot 1`] = `
<div>
  <div
    class="loginPage"
  >
    1
    <h1
      class="loginPage-title"
    >
      Log in to Nodepop
    </h1>
    <form
      class="loginForm"
    >
      <div
        class="formField loginForm-field"
      >
        <label
          class="formField-label"
        >
          <span>
            phone, email or username
          </span>
          <input
            autocomplete="off"
            class="formField-input"
            name="email"
            type="text"
            value=""
          />
        </label>
      </div>
      <div
        class="formField loginForm-field"
      >
        <label
          class="formField-label"
        >
          <span>
            password
          </span>
          <input
            autocomplete="off"
            class="formField-input"
            name="password"
            type="password"
            value=""
          />
        </label>
      </div>
      <input
        name="remember"
        type="checkbox"
        value="remember"
      />
      <select>
        <option
          value="1"
        >
          Option 1
        </option>
        <option
          value="2"
        >
          Option 2
        </option>
        <option
          value="3"
        >
          Option 3
        </option>
      </select>
      <input
        type="file"
      />
      <button
        class="sc-bczRLJ jSwChQ loginForm-submit"
        disabled=""
        type="submit"
      >
        Log in
      </button>
    </form>
  </div>
</div>
`;
